appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/model/util/AnnotationInfo.java

```diff
@@ -0,0 +1,201 @@
+
+/**
+ * Represents the aggregated annotations on a type, its fields and methods including annotations "inherited" from
+ * super-classes and implemented interfaces.
+ */
+public final class AnnotationInfo<T> {
+
+    private static final Map<Class<?>, AnnotationInfo<?>> TYPES = new ConcurrentHashMap<>();
+
+    @SuppressWarnings("unchecked")
+    public static <T> AnnotationInfo<T> valueOf(Class<T> type) {
+        return (AnnotationInfo<T>) TYPES.computeIfAbsent(type, key -> new AnnotationInfo<>(key));
+    }
+
+    private final Class<T> type;
+    private final Map<Class<? extends Annotation>, Annotation> typeAnnotations = new ConcurrentHashMap<>();
+    private final Map<String, Map<Class<? extends Annotation>, Annotation>> fieldAnnotations = new ConcurrentHashMap<>();
+    private final Map<String, Map<Class<? extends Annotation>, Annotation>> methodAnnotations = new ConcurrentHashMap<>();
+    private final Map<String, Map<Class<? extends Annotation>, Annotation>> methodParameterAnnotations = new ConcurrentHashMap<>();
+
+    private AnnotationInfo(Class<T> type) {
+        this.type = type;
+        init(type);
+    }
+
+    public Class<T> getType() {
+        return type;
+    }
+
+    @SuppressWarnings("unchecked")
+    public <A extends Annotation> A getAnnotation(Class<A> annotationType) {
+        return (A) typeAnnotations.get(annotationType);
+    }
+
+    @SuppressWarnings("unchecked")
+    public <A extends Annotation> A getAnnotation(Class<A> annotationType, Field field) {
+        return (A) fieldAnnotations.get(field.getName()).get(annotationType);
+    }
+
+    @SuppressWarnings("unchecked")
+    public <A extends Annotation> A getAnnotation(Class<A> annotationType, Method method) {
+        return (A) methodAnnotations.get(getSignature(method)).get(annotationType);
+    }
+
+    @SuppressWarnings("unchecked")
+    public <A extends Annotation> A getAnnotation(Class<A> annotationType, Parameter parameter) {
+        return (A) methodParameterAnnotations.get(getIdentifier(parameter)).get(annotationType);
+    }
+
+    public <A extends Annotation> A getAnnotation(Class<A> annotationType, AnnotatedElement element) {
+        Class<?> kind = element.getClass();
+        if (kind == Class.class) {
+            return getAnnotation(annotationType);
+        }
+        if (kind == Field.class) {
+            return getAnnotation(annotationType, (Field) element);
+        }
+        if (kind == Method.class) {
+            return getAnnotation(annotationType, (Method) element);
+        }
+        if (kind == Parameter.class) {
+            return getAnnotation(annotationType, (Parameter) element);
+        }
+        return null;
+    }
+
+    public boolean isAnnotationPresent(Class<? extends Annotation> annotationType) {
+        return getAnnotation(annotationType) != null;
+    }
+
+    public boolean isAnnotationPresent(Class<? extends Annotation> annotationType, Field field) {
```

Comment:
These could do with documenting, it's not clear to me from the signature what this method does

Reply:
Added detailed javadoc. It should be clear now.